src/main/java/org/jenkinsci/main/modules/sshd/PublicKeyAuthenticatorImpl.java

```diff
@@ -20,27 +24,40 @@
     private final PublicKeySignatureWriter signatureWriter = new PublicKeySignatureWriter();
 
     public boolean authenticate(String username, PublicKey key, ServerSession session) {
+        boolean authenticated = this.processAuthenticate(username, key);
+
+        if(authenticated){
+            UserDetails userDetails = Jenkins.getInstance().getSecurityRealm().loadUserByUsername(username);
+            SecurityListener.fireAuthenticated(userDetails);
+        }else{
+            SecurityListener.fireFailedToAuthenticate(username);
+        }
+
+        return authenticated;
+    }
+    
+    private boolean processAuthenticate(String username, PublicKey key) {
```

Comment:
maybe it makes sense to return `@CheckForNull User` from this call. But I see there is no way to get `UserDetails` from `User` :(

Reply:
In [#3074](https://github.com/jenkinsci/jenkins/pull/3074) I proposed a new method to retrieve UserDetails for impersonation, but nothing related to regular authentication.